analyzer/aggregate_stats.go

```diff
@@ -79,3 +126,143 @@ func (a *AggregateStatsAnalyzer) writeToDB(ctx context.Context, batch *storage.Q
+			a.logger.Info("updating daily active account stats", "layer", layer)
+
+			// Query the latest indexed block timestamp.
+			latestBlockTs, err := a.latestBlockTs(ctx, layer)
+			if err != nil {
+				a.logger.Error("failed querying latest indexed block timestamp", "layer", layer, "err", err)
+				continue
+			}
+			latestPossibleWindow := floorWindow(latestBlockTs)
+
+			// Start at the latest already computed window, or at the earliest indexed
+			// block if no stats have been computed yet.
+			// Query the end time of the latest computed window.
+			var latestStatsTs time.Time
+			err = a.target.QueryRow(
+				ctx,
+				a.qf.LatestDailyAccountStatsQuery(),
+				layer,
+			).Scan(&latestStatsTs)
+			switch {
+			case err == nil:
+				// Continues below.
+			case errors.Is(pgx.ErrNoRows, err):
+				// No stats yet. Start at the earliest indexed block.
+				var earliestBlockTs *time.Time
+				earliestBlockTs, err = a.earliestBlockTs(ctx, layer)
+				if err != nil {
+					a.logger.Error("failed querying earliest indexed block timestamp", "layer", layer, "err", err)
+					continue
+				}
+				latestStatsTs = floorWindow(earliestBlockTs)
+			default:
+				a.logger.Error("failed querying latest daily accounts stats window", "layer", layer, "err", err)
+				continue
+			}
+
+			// Compute daily unique active accounts for windows until the latest indexed block.
+			batch := &storage.QueryBatch{}
+			for {
+				nextWindow := latestStatsTs.Add(windowStep)
+				if !latestPossibleWindow.After(nextWindow) {
+					// Cannot yet compute stats for next window. Stop.
+					break
+				}
+
+				windowStart := nextWindow.Add(-windowSize)
+				windowEnd := nextWindow
+
+				// Compute active accounts for the provided time window.
+				var aggregateQuery string
+				switch layer {
+				case layerConsensus:
+					aggregateQuery = a.qf.ConsensusActiveAccountsQuery()
+				default:
+					qf := a.qf.NewWithRuntime(layer)
+					aggregateQuery = qf.RuntimeActiveAccountsQuery()
+				}
+				var activeAccounts uint64
+				if err := a.target.QueryRow(
+					ctx,
+					aggregateQuery,
+					windowStart, // from
+					windowEnd,   // to
+				).Scan(&activeAccounts); err != nil {
+					a.logger.Error("failed to compute daily active accounts", "layer", layer, "window_start", windowStart, "window_end", windowEnd, "err", err)
+					continue
+				}
+
+				// Insert computed active accounts.
+				batch.Queue(a.qf.InsertDailyAccountStatsQuery(), layer, windowEnd.UTC(), activeAccounts)
+				if batch.Len() > dailyActiveAccountsBatchLimit {
+					break
+				}
+				latestStatsTs = nextWindow
+			}
+			if err := a.writeToDB(ctx, batch, fmt.Sprintf("update_daily_active_account_%s_stats", layer)); err != nil {
+				a.logger.Error("failed to insert daily active account stats update", "err", err, "layer", layer)
+			}
+			a.logger.Info("updated daily active account stats", "layer", layer, "num_inserts", batch.Len())
+		}
+
+		select {
+		case <-time.After(windowStep):
```

Comment:
Makes sense, although I'm not totally sure about a good tolerance value to use.

In order to compute stats for a window, there needs to be at least one block indexed in the window after (since the analyzer tracks time by looking at timestamps of indexed blocks).
`1s` is surely too small, since consensus blocks are generated every ~6 seconds, therefore we could miss the first block in the new window and not start analyzing the previous window. There is also some lag, so I would use something like 10 seconds.

There are also two cases where the next iteration should be done sooner:
- batch limit reached: the analyzer is way behind latest indexed block, start next iteration sooner
- no stats computed: we likely hit the issue described above (no blocks indexed) retry sooner

---

All in all, I think the best solution is to just use a short constant here (e.g. 1 min), not related to the `windowStep` since the iteration itself is quite lightweight if there's no new data (one latest stats + one latest block query for each layer).

Reply:
:+1: Good thoughts on the tolerance, and the other two cases! I agree with your thinking; a constant wait is probably good enough for out scenario, so keeping it simple is the way to go. We can always tweak/complicate later if needed, not that I expect we'll need to.